ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveOptimizeInlineArrayTableFunctionRule.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.hadoop.hive.ql.optimizer.calcite.HiveRelFactories;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveTableFunctionScan;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This rule optimizes the inline udtf in a HiveTableFunctionScan when it
+ * has an array of structures. The RelNode for a HiveTableFunctionScan places
+ * the input references as the first elements in the return type followed by
+ * the udtf return value which represents the items in the generated table. Take the
+ * case where the base (input) table has col1, and the inline function is represented by:
+ * inline(array( struct1(col2, col3), struct2(col2, col3), struct3(col2, col3), etc...)),
+ * ...and the return value for the table scan node is (col1, col2, col3). In this case,
+ * the same col1 value is joined with the structures within the inline array for the
+ * col2 and col3 values.
+ *
+ * The optimization is to put the "col1" value within the inline array, resulting in
+ * in the new structure:
+ * inline(array(struct1(col1, col2, col3), struct2(col1, col2, col3), ...)
+ * By doing this, we avoid creating a lateral view join operator and a lateral view forward
+ * operator at runtime.
+ */
+public class HiveOptimizeInlineArrayTableFunctionRule extends RelOptRule {
+
+  public static final HiveOptimizeInlineArrayTableFunctionRule INSTANCE =
+          new HiveOptimizeInlineArrayTableFunctionRule(HiveRelFactories.HIVE_BUILDER);
+
+  public HiveOptimizeInlineArrayTableFunctionRule(RelBuilderFactory relBuilderFactory) {
+    super(operand(HiveTableFunctionScan.class, any()), relBuilderFactory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    final HiveTableFunctionScan tableFunctionScanRel = call.rel(0);
+
+    Preconditions.checkState(tableFunctionScanRel.getCall() instanceof RexCall);
+    RexCall udtfCall = (RexCall) tableFunctionScanRel.getCall();
+    if (!udtfCall.getOperator().getName().equalsIgnoreCase("inline")) {
```

Comment:
Yeah, this makes sense, and I usually code things this way, not sure why I did it this way here. 

But it did also make me think:  Should I use the Hive SqlFunctionConverter.reverseOperatorMap.get(SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR) to get the constant?  And if I did that, should I do a null check there to make sure it's valid?  Or should I just keep it the way it is.  

I can't decide which is better. I like that way because it should theoretically be the same constant string. But it does seem a little awkward to code it that way, I would prefer a general constant for "array" within that class but that seems like too much of a rewrite....or maybe I should just do it for this one function?

Thoughts?

Reply:
How about
```
Objects.equals(firstOperand.getOperator().getName(), SqlFunctionConverter.reverseOperatorMap.get(SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR))
```
?
However in this case both `null`s are still considered true which is not ok.

Btw could you please elaborate this? `I would prefer a general constant for "array" within that class`

Reply:
I guess what I'm saying is that I don't like having to use the "reverseOperatorMap" to get it.  I guess what i'm saying is that this line within SqlOperatorConverter:

  registerFunction("array", SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR, hToken(HiveParser.Identifier, "array"));
  
  ...should have been...
  
   registerFunction(ARRAY_FUNCTION, SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR, hToken(HiveParser.Identifier, ARRAY_FUNCTION));
  
  ...which would have avoided the awkward call to reverseOperatorMap.  I guess I can change it for this one constant and that would be easy enough, but all the other strings for registerFunction don't use constants.  
  
  Ok...I'm talking myself into changing it into a constant though.  We can change other registerFunction calls later?

Reply:
Just made the change.  i don't feel so bad about adding it in FunctionRegistry since there is precedence of other constants for function names there.